### ISSUE TITLE
Incorrect validation results

### ISSUE THREAD

**Issue by reuteran** [rtrlib/rtrlib#99]
For prefix 1.9.23.0/24 there are two ROAs:
```
ASN: 4788
Prefix: 1.9.0.0/16
Max length: 24
```

```
ASN: 65120
Prefix: 1.9.23.0/24
Max length: 24
```

I used `'pfx_table_validate_r(..)' ` to validate this:
```
ASN: 4788
Prefix: 1.9.23.0/24
```

And the result is `BGP_PFXV_STATE_INVALID`, but it should be `BGP_PFXV_STATE_VALID`.

I've confirmed this using the cli-validator:

```
rtrlib/tools $ ./cli-validator rpki-validator.realmv6.org 8282
Arguments required: IP Mask ASN
1.9.23.0 24 4788 //Input
1.9.23.0 24 4788|65120 1.9.23.0 24 24|2 //Output, 2 means INVALID
1.9.23.0 24 65120 //Input
1.9.23.0 24 65120|65120 1.9.23.0 24 24|0 //Output, 0 means VALID
```

I checked the ROAs at http://rpki-validator.realmv6.org:9090/roas (search for 1.9.23.0/24)

It looks like the validation code is only checking the ROA with the most specific prefix match, and if that ASN doesn't match it says INVALID.

**Comment by waehlisch** [comment]
You are right. That's an embarrassing bug. Can someone check the code and fix this ASAP?

@reuteran did you check that the /16 ROA is part of the pfx_table?

Actually, if you are right, the current code searches for the most specific covering (not matching) ROA.

**Comment by reuteran** [comment]
@waehlisch 

> did you check that the /16 ROA is part of the pfx_table?

Not directly, since gdb won't let me step into it for some reason (compiled with debug symbols).
But I queried for:
```
asn: 4788
prefix: 1.9.0.0/16
```
which returns:
```
1.9.0.0 16 4788|4788 1.9.0.0 16 24|0
```
The section after the first '|' means that there is indeed an pfx_table entry for 1.9.0.0/16-24 for AS4788.

**Comment by waehlisch** [comment]
OK. Still embarrassing. Can you fix this?

**Comment by reuteran** [comment]
OK, but it might take a bit because I am unfamiliar with the pfx_table. Hopefully I'll have it by tomorrow

**Comment by smlng** [comment]
I added a test in #101 that should reflect the described case, however the test passes successful not showing the same problem.

**Comment by waehlisch** [comment]
Looks correct to me. But I'm in a rush. @reuteran, can you check?

**Comment by reuteran** [comment]
The test doesn't reproduce the problem because the length of the shortest prefix is much longer than the height of the prefix tree.
I think this example should show the bug:
Add to empty tree:
```
   255.0.0.0 min_len: 26, max_len: 26, AS A
   255.0.0.0 min_len: 25, max_len: 25, AS B
   255.0.0.0 min_len: 24, max_len: 24, AS C
   128.0.0.0 min_len: 1, max_len: 24, AS D
```
Then validate with prefix=255.0.0.0/24, asn=D. This should give invalid.

**Comment by smlng** [comment]
@reuteran interesting, your example already works by adding the last 2 lines, only. This should give valid, but the test fails ... 

Are you working on a fix for rtrlib?

**Comment by waehlisch** [comment]
We don't see an easy fix of this problem. If you have solution, please share.
@reuteran will work on Trie, as an alternative data structure.

cc @tcschmidt

**Comment by fho** [comment]
Can this issue be closed? It seems to have been fixed by #103?

**Comment by fho** [comment]
https://github.com/rtrlib/rtrlib/issues/99#issuecomment-261752485
> I think this example should show the bug:
> Add to empty tree:
> 
>    255.0.0.0 min_len: 26, max_len: 26, AS A
>    255.0.0.0 min_len: 25, max_len: 25, AS B
>    255.0.0.0 min_len: 24, max_len: 24, AS C
>    128.0.0.0 min_len: 1, max_len: 24, AS D
> Then validate with prefix=255.0.0.0/24, asn=D. This should give invalid.

I wonder why 255.0.0.0/24, asn=D should give an invalid validation result.

AS D is authorized to announce routes for 128.0.0.0/1-24.
Prefix length 1 means, that only the first bit of the IP-address determines the network.
Therefore valid host addresses of 128.0.0.0/1 of AS D are  128.0.0.1-255.255.255.254.
The network 255.0.0.0/24 is part of this range.
Should'nt be the validation result for 255.0.0.0/24, ASN D be valid?

AS C has a more specific ROA for the network but the RFC states:

https://tools.ietf.org/html/rfc6483#page-3
> In an environment of a collection of valid ROAs, a route's validity
> state is considered to be "valid" if any ROA provides a "valid"
> outcome.

**Comment by smlng** [comment]
@fho you're right about AS D and that is authorized to announce the prefix. However, RTRlib with lpfst will resolve to invalid (which is wrong), that's what @reuteran is pointing out here, i.e. he describes the failure case *not* the expected validation result!

**Comment by fho** [comment]
@smlng thanks!

Do I understand it right that the problem with the current implementation is the following?:

The LPFST looks like this:
```

          10.100.255.0/24-24
              /                   \
             /                     \
128.0.0.0/1-24           255.0.0.0/24-24
```
We are trying to verify 255.0.0.0/24. 
The lpfst_lookup() starts searching at the root node, compares bit 0 of 10.100.255.0 (1100) and 255.0.0.0 (1111, they don't match.
Because bit 0 of the searched prefix is 1 it continues the search at the left child node and never find the matching pfx record at the right child node.

**Comment by reuteran** [comment]
> Because bit 0 of the searched prefix is 1 it continues the search at the left child node and never find the matching pfx record at the right child node.

The other way I around I think. If the bit is 1, it goes into the right tree. So in your example, it would never search the 128.0.0.0/1-24 record.

**Comment by smlng** [comment]
Its a bit complicated, because the lpfst search in itself is successful -> it finds the node with `255.0.0.0/24` but the ROA contained there is with AS A, which gives an invalid as we initiated our search with `255.0.0.0 24 AS-D`. However, there is a valid ROA in node `128.0.0.0/1` with max length 24 and AS D. Which is never found with lpfst, and never will be because of its bit-wise search.

I tried to fix the lpfst implementation, but without success (see #104) and IMO lpfst is essentially not suitable for ROAs, at least not with the current search conditions.

**Comment by fho** [comment]
Yes, the lpfst seems to not be able to handlle that scenario correctly.
I saw #104, #103 but it takes a lot of time for me to get the ideas by only reading the code without any textual descriptions.

I tried to convert the LPFST to a shortest prefix search tree, only 1 line of code needs to be changed in the lpfst for it.
The discussed scenario worked. A few test failed in test_lpfst.c but the ones I verified  failed because a matching /16 record was returned instead of an expected matching /24 record.
Any objections against a spfst datastructure? (is it maybe the same idea than https://github.com/rtrlib/rtrlib/pull/103?)

**Comment by smlng** [comment]
@fho I tried to mod/fix the lpfst, too - but without success, I found ways to fix some of the failed validation results but other still persisted. 

If you have an alternative solution, don't hesitate to make a PR. @ColinBS has a test framework with more extensive offline tests to verify and compare proposed fixes.

**Comment by waehlisch** [comment]
@reuteran and I also thought about a fix. But actually there is no simple solution. Very most likely a potential fix will result in O(number of prefixes) search complexity or alternatively in significantly increased memory consumption. The problem is that you don't easily know which sub-tree you should traverse to find _all_ ROAs.

I don't think that a fix of LPFST will bring better performance compared to a Trie.

**Comment by fho** [comment]
My solution can be found at https://github.com/rtrlib/rtrlib/tree/issue/99.
I  changed the ordering of the nodes in LPFST to sort nodes with shorter prefixes before nodes with longer prefixes. 
The whole fix in the lpfst are just 2 lines: https://github.com/rtrlib/rtrlib/commit/d85c21ca01ce26ce3cbc5619e5951954c9696239
I also had to rework the lpfst tests because of the changed tree structure, all tests are successful.

It seems almost the same to the solution than the one from @reuteran, is it? :-) 
If yes, we could also go with @reuteran solution after the commit messages in his branch were reworked.

Any objections?
Outstanding in my branch is renaming *pfst to tree or so and reworking the test_pfx, test_issue99 files.

I tried to verify the difference with the rtr-testsuite but it failed because of https://github.com/rtrlib/rtr-testsuite/issues/1.

**Comment by reuteran** [comment]
@fho Your fix is essentially the same as I did. 
However the 'lpfst_lookup' function doc says:
> Searches for the node with the longest prefix, matching the passed ip prefix and prefix length.

So I rewrote it for the trie so it still returns the node with the longest prefix. Its not needed anywhere, because validation wants to get the first matching prefix in the tree (which used to be the longest in the lpfst). So right now its just being called by tests.

I replaced 'lpfst_lookup' with 'trie_lookup_first_match'. I'll add an explanation of the trie to the PR, but its essentially just a inverted version of the lpfst (so shortest prefix first search tree).

**Comment by fho** [comment]
@reuteran Ah ok :-)
I'm fine with using your version if you ant to adapt it.
Sounds better for me to change the description of lpfst_lookup() and remove trie_lookup_first_match().
Maybe you then want to cherry-pick my changed lpfst test, that works with the spfst: https://github.com/rtrlib/rtrlib/commit/1261e3bca55cb3da44bd45eaa8a6e6b8b048e663

https://github.com/rtrlib/rtrlib/commit/ed9c5bb613e129f4b38a8d757aba3f560f4f70dd could also be worth to cherry-pick to your branch

**Comment by waehlisch** [comment]
@fho, I'm heavily confused. Trie is a different data structure compared to LPFST. As such we should also have a different API. Keeping the name lpfst_lookup()  to operate on a Trie may lead to incorrect assumptions (regarding the semantic of the function) at the programmer side.

**Comment by fho** [comment]
@waehlisch Yes, the functions have to be renamed.
What I tried to say was: keep the old function naming scheme in @reuterans branch, rename lpfst_lookup() to trie_lookup(). trie_lookup() should return the first matching node in the tree (like before).
The function trie_lookup_first_match() can then be removed.

**Comment by fho** [comment]
@reuteran do you think you'll be able to finish the pull request until Monday?
Otherwise I suggest, that I'll finish my version of the fix and ask for a review with a pull request for it.

We should make progess with this, it's already open since 2.5 month and it's an major issue that makes the library unusable.

**Comment by smlng** [comment]
I'll check with @ColinBS, he has some tool chain to verify and compare validation results of the different solutions.

**Comment by waehlisch** [comment]
@smlng yes, before this test has not been completed, we will not merge any code in this context.

**Comment by ColinBS** [comment]
I improved the script to print the exact ROAs with the status of lpfst, trie and the RIPE RPKI validator.
[statistics.txt](https://github.com/rtrlib/rtrlib/files/754476/statistics.txt)

**Comment by fho** [comment]
@ColinBS cool, which commits do you used exactly?
88f49f1 for trie? 16879b3 for lpfst?

**Comment by fho** [comment]
There are 134 validations where RIPE RPKI validator result is not found and rtrlib+trie validates 133x valid + 1x invalid
(awk '{if ($6 != $8)  print $0 }' gistfile1.txt)

**Comment by ColinBS** [comment]
@fho yes, these are the commits I use. I will double check the problem. There might be a mistake on my side.

**Comment by waehlisch** [comment]
@ColinBS can you please add your scripts to rtrlib/tools (feel free to add a subdirectory).

**Comment by ColinBS** [comment]
I added the script files to the tools repo.
Here is the result from the spfst implementation:
https://gist.github.com/ColinBS/529458b379bc560d681cafdda090b57d

**Comment by waehlisch** [comment]
@ColinBS perfect; nice scripting! Would be great if you add the RIPEVal back again. Then we should have everything in place.

**Comment by fho** [comment]
Would be awesome to integrate it in travis, so that on every git-push rtrlib results would be validated against the ones from the rpki-validator automatically.
But that should we put in a new github-issue :-)

**Comment by waehlisch** [comment]
Fixed in PR #112.
